### PR TITLE
Validate blank field name

### LINKED ISSUE #19251: Blank field names should be illegal
Today you can create a blank field name, eg:

    PUT t/t/1
    {
      "": 5
    }
    
Or

    PUT t/t/1
    {
      "": {
        "foo": 5
      }
    }
    

Blank field names should be illegal as they are not accessible via paths

Related https://github.com/elastic/kibana/issues/7617

### PR DESCRIPTION
**Pull request by johtani**
Add validation to Mapper constructor.
Add test cases
Add rest-api-test

Add only Mapper without Legacy*Mapper
And we should discuss how to handle already exist indices that has blank field name when it is restored

Closes #19251

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/index/mapper/Mapper.java [hunk 1]
```diff
@@ -172,6 +172,9 @@ public ParserContext createMultiFieldContext(ParserContext in) {
     private final String simpleName;
 
     public Mapper(String simpleName) {
+        if (Strings.isNullOrEmpty(simpleName)) {
+            throw new IllegalArgumentException("name cannot be null and empty string");
+        }
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/index/mapper/Mapper.java]: I would distinguish the null check from the empty check, as a null simpleName means that we have a bug while an empty simpleName means the issue is on the user end. So maybe something like this:
```
Objects.requireNonNull(simpleName);
if (simpleName.isEmpty()) {
    throw new IllegalArgumentException("name cannot be an empty string");
}
```

**Comment by jpountz** [comment]
> And we should discuss how to handle already exist indices that has blank field name when it is restored

Maybe we should only enforce this on 5.0+ indices?

**Comment by johtani** [comment]
@jpountz Thanks for your comment.
If only enforce this on 5.0+ indices, the validation should be moved to other place, maybe Builder.build method or something else, right?

**Comment by johtani** [comment]
Hi @jpountz ,
I moved the check logic to Mapper.Builder.build method and validate only index created version is 5.x.
I have a question about BuilderContext.
I'm not sure BuilderContext doesn't have the IndexSettings and some test case uses Settings.Empty.
Is it the right place to validate field names with version?

File: core/src/main/java/org/elasticsearch/index/mapper/Mapper.java [hunk 2]
```diff
@@ -55,7 +54,7 @@ public Settings indexSettings() {
 
         @Nullable
         public Version indexCreatedVersion() {
-            if (indexSettings == null) {
+            if (indexSettings == null || indexSettings.isEmpty()) {
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/index/mapper/Mapper.java]: let's not add leniency to this method, why did you need to add it?

File: core/src/main/java/org/elasticsearch/index/mapper/Mapper.java [hunk 3]
```diff
@@ -77,7 +76,17 @@ public String name() {
         }
 
         /** Returns a newly built mapper. */
-        public abstract Y build(BuilderContext context);
+        public final Y build(BuilderContext context) {
+            if (context.indexCreatedVersion() != null && context.indexCreatedVersion().onOrAfter(Version.V_5_0_0_alpha1)) {
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/index/mapper/Mapper.java]: why do we need to check that `context.indexCreatedVersion()` is not null? the version should always be available

**Comment by jpountz** [comment]
> I'm not sure BuilderContext doesn't have the IndexSettings and some test case uses Settings.Empty.

This looks like a bug indeed!

> Is it the right place to validate field names with version?

I don't like how this requires to add a buildMapper helper method. Maybe we could move the check to the FieldMapper() and ObjectMapper() constructors?

**Comment by johtani** [comment]
OK. I will move the check logic to each constructors.

File: core/src/main/java/org/elasticsearch/index/mapper/Mapper.java (hunk 3 above)
Comment (johtani) [inline core/src/main/java/org/elasticsearch/index/mapper/Mapper.java]: Because BuilderContext.indexCreatedVersion() has Nullable annotation...
Should we remove Nullable annotation in Mapper.BuilderContext?

Comment (jpountz) [inline core/src/main/java/org/elasticsearch/index/mapper/Mapper.java]: Yes I think we should (but this belongs to a different change)

**Comment by jpountz** [comment]
>  I will move the check logic to each constructors.

Just to be sure we are on the same page: I think this can be done only in the two base classes FieldMapper and ObjectMapper.

**Comment by johtani** [comment]
Move the check logic to both constructors. 
Rebase master and  squash commit because I revert adding Mapper.Builder.buildMapper() method.

**Comment by johtani** [comment]
@jpountz Can you review again?
And should I leave the comment about Nullable or create an issue about it?

File: core/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java [hunk 4]
```diff
@@ -246,6 +246,11 @@ protected FieldMapper(String simpleName, MappedFieldType fieldType, MappedFieldT
         super(simpleName);
         assert indexSettings != null;
         this.indexCreatedVersion = Version.indexCreated(indexSettings);
+        if (indexCreatedVersion.onOrAfter(Version.V_5_0_0_alpha5)) {
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java]: this should be alpha6 now that alpha5 is out

**Comment by jpountz** [comment]
LGTM, please just fix the version that is checked before pushing

**Comment by jpountz** [comment]
> And should I leave the comment about Nullable or create an issue about it?

Let's have a separate issue.

**Comment by jpountz** [comment]
LGTM

**Comment by johtani** [comment]
@jpountz Thanks! I made #20174 just now. Oops, I have to rebase this PR again for fixing CI fail...
